File: report/src/main/java/com/vmlens/report/element/RunElement.java
package com.vmlens.report.element;

import com.vmlens.report.runelementtype.RunElementType;

public class RunElement {

    private final LoopRunAndThreadIndex loopRunAndThreadIndex;
    private final int runPosition;
    private final StacktraceLeaf stacktraceLeaf;
    private final RunElementType operationTextFactory;

    private final int inMethodId;

    public RunElement(LoopRunAndThreadIndex loopRunAndThreadIndex, int runPosition, StacktraceLeaf stacktraceLeaf,
                      RunElementType operationTextFactory, int inMethodId) {
        this.loopRunAndThreadIndex = loopRunAndThreadIndex;
        this.runPosition = runPosition;
        this.stacktraceLeaf = stacktraceLeaf;
        this.operationTextFactory = operationTextFactory;
        this.inMethodId = inMethodId;
    }

    public int runId() {
        return loopRunAndThreadIndex.runId();
    }

    public LoopRunAndThreadIndex loopRunAndThreadIndex() {
        return loopRunAndThreadIndex;
    }

    public int runPosition() {
        return runPosition;
    }

    public StacktraceLeaf stacktraceLeaf() {
        return stacktraceLeaf;
    }

    public RunElementType operationTextFactory() {
        return operationTextFactory;
    }

    public int inMethodId() {
        return inMethodId;
    }
}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/RunElement.scala
package com.anarsoft.race.detection.report.element

import com.anarsoft.race.detection.report.element.runelementtype.ReportOperation


class RunElement(val loopRunAndThreadIndex: LoopRunAndThreadIndex, 
                 val runPosition: Int, 
                 val stacktraceLeaf: StacktraceLeaf,
                 val operationTextFactory: ReportOperation,
                 val inMethodId: Int) extends  Ordered[RunElement] {


  override def compare(that: RunElement): Int = {
    if(loopRunAndThreadIndex.runId != that.loopRunAndThreadIndex.runId) {
      loopRunAndThreadIndex.runId.compare(that.loopRunAndThreadIndex.runId)
    } else {
      runPosition.compare(that.runPosition)
    }
  }


  private def canEqual(other: Any): Boolean = other.isInstanceOf[RunElement]
  
  override def equals(other: Any): Boolean = other match {
    case that: RunElement =>
      that.canEqual(this) &&
        loopRunAndThreadIndex == that.loopRunAndThreadIndex &&
        stacktraceLeaf == that.stacktraceLeaf &&
        operationTextFactory == that.operationTextFactory &&
        inMethodId == that.inMethodId
    case _ => false
  }
  
  override def hashCode(): Int = {
    val state = Seq(loopRunAndThreadIndex, stacktraceLeaf, operationTextFactory, inMethodId)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
